### ISSUE TITLE
Tracking Issue for RFC #2972: Constrained Naked Functions 

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#90957]
This is a tracking issue for the RFC "Constrained Naked Functions" (rust-lang/rfcs#2972).
The feature gate for the issue is `#![feature(naked_functions)]`.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [x] Implement the RFC
- [x] Confirm that all errors and warnings are emitted properly
- [ ] Final review of attributes that are incompatible with `#[naked]`
- [ ] Determine precise inlining gurantees
- [x] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide])
- [x] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])
  - https://github.com/rust-lang/rust/pull/134213

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Unresolved Questions

None.

**Comment by safinaskar** [comment]
"produce a clear warning if any of the above suggestions are not heeded" in RFC seems to contain typo

**Comment by bstrie** [comment]
This is implemented, checking the box.

Adding a new step: "Confirm that all errors and warnings are emitted properly".

**Comment by bstrie** [comment]
# Request for Stabilization

A proposal that the `naked_functions` feature be stabilized for Rust 1.60, to be released 2022-04-07. A stabilization PR can be found at https://github.com/rust-lang/rust/pull/93587 .

## Summary

Adds a new attribute, `#[naked]`, which may be applied to functions. When applied to a function, the compiler will not emit a [function prologue](https://en.wikipedia.org/wiki/Function_prologue_and_epilogue) when generating code for this function. This attribute is analogous to [`__attribute__((naked))`](https://developer.arm.com/documentation/100067/0608/Compiler-specific-Function--Variable--and-Type-Attributes/--attribute----naked---function-attribute) in C. The use of this feature allows the programmer to have precise control over the assembly that is generated for a given function.

The body of a naked function must consist of a single `asm!` invocation. This `asm!` invocation is heavily restricted: the only legal operands are `const` and `sym`, and the only legal options are `noreturn` (which is mandatory) and `att_syntax`. In lieu of specifying operands, the `asm!` within a naked function relies on the specified `extern` calling convention in order to determine the validity of registers.

An example of a naked function:

```rust
const THREE: usize = 3;

#[naked]
/// Adds three to a number and returns the result.
pub extern "sysv64" fn add_n(number: usize) -> usize {
    // SAFETY: the validity of these registers is guaranteed according to the "sysv64" ABI
    unsafe {
        asm!(
            "add rdi, {}"
            "mov rax, rdi"
            "ret",
            const THREE,
            options(noreturn)
        );
    }
}
```

## Documentation

The Rust Reference: https://github.com/rust-lang/reference/pull/1153

## Tests

* [codegen/naked-functions.rs](https://github.com/rust-lang/rust/blob/dca1e7aa5a8ac05ddaea731f4eab20de91acb46b/src/test/codegen/naked-functions.rs): tests for the absence of a function prologue.
* [codegen/naked-noinline.rs](https://github.com/rust-lang/rust/blob/dca1e7aa5a8ac05ddaea731f4eab20de91acb46b/src/test/codegen/naked-noinline.rs): tests for the presence of `naked` and `noinline` LLVM attributes.
* [ui/asm/naked-functions.rs](https://github.com/rust-lang/rust/blob/dca1e7aa5a8ac05ddaea731f4eab20de91acb46b/src/test/ui/asm/naked-functions.rs): tests that naked functions cannot accept patterns as function parameters, that referencing their function parameters is prohibited, that their body must consist of a single `asm!` block, that they must specify the `noreturn` option, that they must not specify any other option except `att_syntax`, that they only support `const` and `sym` operands, that they warn when used with `extern "Rust"`, and that they are incompatible with the `inline` attribute.
* [ui/asm/naked-functions-ffi.rs](https://github.com/rust-lang/rust/blob/dca1e7aa5a8ac05ddaea731f4eab20de91acb46b/src/test/ui/asm/naked-functions-ffi.rs): tests that a warning occurs when the function signature contains types that are not FFI-safe.
* [ui/asm/naked-functions-unused.rs](https://github.com/rust-lang/rust/blob/dca1e7aa5a8ac05ddaea731f4eab20de91acb46b/src/test/ui/asm/naked-functions-unused.rs): tests that the `unused_variables` lint is suppressed within naked functions.
* [ui/asm/naked-invalid-attr.rs](https://github.com/rust-lang/rust/blob/dca1e7aa5a8ac05ddaea731f4eab20de91acb46b/src/test/ui/asm/naked-invalid-attr.rs): tests that the naked attribute is only valid on function definitions.

## History

This feature was originally proposed in [RFC 1201](https://github.com/rust-lang/rfcs/pull/1201), filed on 2015-07-10 and accepted on 2016-03-21. Support for this feature was added in #32410, landing on 2016-03-23. Development languished for several years as it was realized that the semantics given in RFC 1201 were insufficiently specific. To address this, a minimal subset of naked functions was specified by [RFC 2972](https://github.com/rust-lang/rfcs/pull/2972), filed on 2020-08-07 and accepted on 2021-11-16. Prior to the acceptance of RFC 2972, all of the stricter behavior specified by RFC 2972 was implemented as a series of warn-by-default lints that would trigger on existing uses of the `naked` attribute; these lints became hard errors in #93153 on 2022-01-22. As a result, today RFC 2972 has completely superseded RFC 1201 in describing the semantics of the `naked` attribute.

## Unresolved Questions

* In C, `__attribute__((naked))` prevents the compiler from generating both a function prologue and a function epilogue. However in Rust it was [discovered](https://github.com/rust-lang/rust/issues/32408#issuecomment-784037555) that under some circumstances LLVM will choose to emit instructions following the body of a naked function, which is seemingly non-trivial to prevent. Since most of the utility of naked functions comes from preventing the prologue rather than the epilogue, for the time being this feature has restricted itself to only guaranteeing the absence of the prologue. Instead, the reference will specify that users must not rely on the presence of code following the body of a naked function, and that a future version of Rust reserves the right to guarantee the absence of such code.
* Prior to stabilization, [concerns were raised](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/is.20inline.28never.29.20guaranteed.3F/near/270032604) that the current implementation might not be capable of guaranteeing that a naked function is never inlined in every circumstance. Obviously the correctness of naked functions relies on Rust setting up the callstack as if a function call were being performed, even in cases where the function is ultimately inlined. However, there may be observable differences in behavior based on whether or not a naked function is inlined, e.g. due to exported symbols or named labels. Thus the reference will specify that implementations are not currently required to prevent the inlining of naked functions, but that a future version of Rust reserves the right to guarantee that naked functions are not inlined, and that users must not rely on any observable differences that may arise due to the inlining of a naked function.

**Comment by bjorn3** [comment]
> Since most of the utility of naked functions comes from preventing the prologue rather than the epilogue, for the time being this feature has restricted itself to only guaranteeing the absence of the prologue.

There is no observable difference between LLVM adding an epilogue and a following unnamed function starting with instructions identical to a regular epilogue.

**Comment by bstrie** [comment]
@bjorn3 I believe it can cause problems in circumstances such as https://github.com/rust-lang/rust/issues/32408#issuecomment-898815309 . If someone were to work around that issue by assuming the existence of extra instructions and manually accounting for it, then their code would be broken if Rust ever stopped generating those instructions. This is the sort of edge case that the cautious language is designed to address.

**Comment by roblabla** [comment]
> > Since most of the utility of naked functions comes from preventing the prologue rather than the epilogue, for the time being this feature has restricted itself to only guaranteeing the absence of the prologue.
> 
> There is no observable difference between LLVM adding an epilogue and a following unnamed function starting with instructions identical to a regular epilogue.

I believe it's possible to observe a difference when `naked` is coupled with `#[link_section]`. You could apply that attributed to a naked function to put it in a well-known section that you expect to only contain your function.

**Comment by joshtriplett** [comment]
:+1: for stabilizing, though I'd like to make sure that there's consensus on https://github.com/rust-lang/rust/issues/32408#issuecomment-1016931998 .

**Comment by joshtriplett** [comment]
Shall we stabilize constrained naked functions?

@rfcbot merge

**Comment by Amanieu** [comment]
I'd like to have a whitelist of attributes that are allowed to be applies to a naked function. In the future we may want to lower naked functions directly to LLVM module-level assembly instead of using LLVM's native naked function support. For that to happen rustc needs to be able to manually translate any attributes (e.g. `#[link_section]`) to the corresponding asm directive.

**Comment by comex** [comment]
> In the future we may want to lower naked functions directly to LLVM module-level assembly instead of using LLVM's native naked function support.

Why, because of the epilogue issue?  Personally I would much rather see that fixed on LLVM’s end, rather than going to great lengths to work around it in rustc.  Or is there another reason?

**Comment by bjorn3** [comment]
Lowering to `global_asm!` would allow all codegen backends to share the same naked function handling. Also codegen backends without native inline assembly support (like cg_clif) will have to lower to `global_asm!` even if no code is shared between codegen backends, so it will need this same restriction.

**Comment by bstrie** [comment]
@comex In addition to the epilogue issue, it seems the inlining issue could also be resolved by lowering to `global_asm!`.

@Amanieu I'd be interested in seeing such a list. Obviously part of the appeal of naked functions is that they can be treated just like a normal function in most cases, which includes the ability to be marked with attributes (e.g. `#[doc]`), so I'm curious how restrictive this list would be. It occurs to me that, in addition to forbidding `#[inline]`, naked functions also already forbid `#[track_caller]`; I should probably mention this in the reference.

**Comment by roblabla** [comment]
Most attributes that work with arbitrary functions should work with naked functions IMO. Taking the list [here](https://doc.rust-lang.org/reference/attributes.html), those make sense to use with naked fns:

- Conditional compilation: `cfg`, `cfg_attr`
- Diagnostics: `allow` and company, `deprecated`
- ABI...: `link_section`, `export_name`, `no_mangle`, `used`
- Documentation: `doc`

The only ones that I think make sense to disallow are those in the "Code Generation" category: `inline`, `cold`, `track_caller`, `target_feature`

**Comment by joshtriplett** [comment]
@roblabla I think `cold` makes sense on a naked function. It may affect where the code ends up in the final binary, and what it's placed next to, all of which seems fine.

I think `target_feature` makes a *little* sense, if you want to use it for its semantic of "you shouldn't reference this function outside of an unsafe block or a function with the same target feature", but it'd take some care to get right; prohibiting `target_feature` on a naked function seems fine for now until we nail down that semantic.

I agree that `inline` and `track_caller` don't make sense on a naked function.

**Comment by nikomatsakis** [comment]
@joshtriplett 

>  And must_use doesn't make sense, since the function doesn't actually have a normal return value.

Can't naked functions declare a return value if they wish? it's just up to the caller to manage the ABI side of things.

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

@bstrie thank you for the summary! It answered my main questions.

**Comment by npmccallum** [comment]
I'm just adding a note as the author of Constrained Naked Functions that I support this direction and I believe all substantial issues are resolved. After this, we should be in a position to improve functionality incrementally as needed in a well defined way. 👍 🚀 🥳

**Comment by fintelia** [comment]
> Obviously the correctness of naked functions relies on Rust setting up the callstack as if a function call were being performed, even in cases where the function is ultimately inlined.

Are you sure that this actually happens when rustc decides to inline a naked function? In the past, [inlining a naked function also skipped setting up the callstack properly](https://github.com/rust-lang/rfcs/pull/2774#issuecomment-536145700).

> However, there may be observable differences in behavior based on whether or not a naked function is inlined, e.g. due to exported symbols or named labels.

If I understand correctly, this is equivalent to saying that **asm blocks in naked functions may not export symbols** because you'd get linker errors if an inlined and non-inlined version were both included in the binary?

**Comment by bstrie** [comment]
> Are you sure that this actually happens when rustc decides to inline a naked function? In the past, https://github.com/rust-lang/rfcs/pull/2774#issuecomment-536145700.

We're at the mercy of LLVM to not inline a naked function. We emit LLVM's `noinline` attributes on naked functions, which is documented like so: *"This attribute indicates that the inliner should never inline this function in any situation."* We have (a codegen test)[https://github.com/rust-lang/rust/blob/dca1e7aa5a8ac05ddaea731f4eab20de91acb46b/src/test/codegen/naked-noinline.rs]to ensure that this attribute is emitted properly on naked functions. To my mind that would appear to settle the issue, but some others are concerned that it may not be absolutely bulletproof, so the defensive language you're referring to, while likely overly conservative, is designed to sidestep the need to precisely define "inlining" while still give users assurance that naked functions will behave as they expect even in the absence of that definition. If we update your linked example we can observe that indeed the function is no longer inlined: https://rust.godbolt.org/z/rvPe4r8no , but if you can come up with a counterexample I'd be happy to see it. Personally I think this is a bit of a quibble (Rust has stabilized features with less reference-level rigor than this before), but I'd rather avoid arguing if it means unblocking this feature from stabilizing.

> If I understand correctly, this is equivalent to saying that asm blocks in naked functions may not export symbols because you'd get linker errors if an inlined and non-inlined version were both included in the binary?

I don't believe so, asm blocks already have this issue (any function containing an asm block might be inlined), and even if you use global_asm that doesn't stop an end-user from including two different versions of your crate in their binary (consider LTO as well). Naked functions don't add any new hazard here, if anything they might eventually make it a bit easier since you might be able to leverage the compiler's own mangling scheme to guarantee the uniqueness of global symbols.

**Comment by fintelia** [comment]
I'd prefer that inlining a naked function was always considered a compiler bug. No one wants it to happen. No one has demonstrated a case where it currently happens. No one has shown that the code exists to configure the callstack properly if it did. And at least one case where inlining used to happen _didn't_ do anything to configure the callstack.

The current text is even rather vague about what it means to configure the callstack and registers properly. At a minimum, the program counter register might not equal the function address. But beyond that, some of the Zulip comments suggest that backtraces from inlined functions might not work as expected. Does that mean that the frame pointer (if the ABI specifies one) might hold a misleading value?

The guidance seems to be that if you don't want your function to break when it is inlined, you shouldn't count on "observable differences in behavior". Except that's a circular definition! If you want your function's behavior not to change when inlined, you must make sure your code doesn't behave differently when inlined. 

> The legalese here is designed to tell users that, for the moment, any number of copies of the function might exist in the binary, but that in the future the compiler may decide not to include that many copies, and that if you do any bizarre thing that somehow relies on there being more than one copy in the binary then you are outside of the stability guarantee; I find it implausible that anyone could ever write reasonable code that would run afoul of this clause.

The case I'm worried about is that the compiler currently inserts exactly one copy of the function and everything works fine, but without warning later breaking things by deciding to include multiple copies. At worst, the old and new versions might both compile without errors or warnings but behave differently at runtime.

**Comment by bstrie** [comment]
@fintelia I'm happy to tweak the proposed language in the reference (PRs accepted). As mentioned, I think the risk of anything going wrong here is quite low, and plenty of things to do with soon-to-be-stable-asm already feature relatively fragile semantics (e.g. [labels](https://github.com/rust-lang/rust/issues/81088) and [directives](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/asm!.20and.20backends/near/270189346)). However at this point what matters most is 1) what semantics the language team is willing to commit to, and 2) what semantics the compiler team thinks that rustc is reasonably capable of enforcing. I've added a new checkbox for this concern. I encourage you to resurrect that Zulip thread and raise your concerns there as well.

**Comment by archshift** [comment]
This was recently marked ready to merge 3 months ago. Has anything changed since then?

**Comment by joshtriplett** [comment]
I think it just needs a stabilization PR.

**Comment by Lokathor** [comment]
The RFC says that the `asm!` "must not contain any other options except `att_syntax`", but the `raw` option should also be permitted.

**Comment by Lokathor** [comment]
Also, having people put `noreturn` on a thing that generally does return is very confusing design, but the FCP for that is over so I don't expect a bigger update like that.

**Comment by joshtriplett** [comment]
@Lokathor Both of those can still change after stabilizing naked functions; that's not a blocker.

@bstrie or others: gentle ping on stabilization?

**Comment by Lokathor** [comment]
update: seems that `raw` is already accepted as an option to naked function inline assembly, so the docs just need to be fixed whenever.

**Comment by repnop** [comment]
while talking to @memoryruins about another feature which might interact poorly with LLVM's function deduplication pass (`fn_align`), it looks like LLVM gladly deduplicates `#[naked]` functions when the body is the same even in the face of different alignment requirements (which appears to pick the higher alignment between the two at least) on the functions: https://rust.godbolt.org/z/WzWTEqM9d

I can't immediately think of how this would cause issues off the top of my head, and it seems the moment that the asm blocks are not _exactly_ equivalent (if you add a comment, for instance) it won't dedup, but seeing as how there were issues with #54685 I wanted to bring attention to it. cc @RalfJung since you pioneered the fix for the original issue and might have some insight I don't here

my immediate feeling here is that non-inlineable functions shouldn't be be able to be deduplicated

**Comment by fintelia** [comment]
@repnop If you do binary patching, then the deduplication you're describing will break things. One example: a program that creates two identical naked functions `bar` and `baz` and then at runtime overwrites the contents of `bar`. Any calls to `baz` are unexpectedly going to have the new behavior.

**Comment by RalfJung** [comment]
https://github.com/rust-lang/rust/issues/54685 was different, that was a case of inconsistent compilation in the face of function deduplication. This is a case of potentially undesirable deduplication. I have no idea what the intended behavior is here (I leave that to the domain experts), nor how to control LLVMs deduplication (I leave that to the LLVM experts).

@fintelia that argument applies equally to non-naked functions, doesn't it? Cc @Amanieu who IIRC said that binary patching is an intended usecase for `asm!`. Should all functions containing `asm!` be exempt from deduplication? (No idea if that is feasible, but that is a separate question.)

**Comment by Amanieu** [comment]
Binary patching is allowed, but the compiler is still allowed to duplicate an `asm!` block. The way that would work is that, from within the `asm!`, you record instruction addresses in a separate section which you can later use to find patchable code blocks.

**Comment by joshtriplett** [comment]
We reviewed this in today's @rust-lang/lang meeting. We'd love to see this stabilized, and we think it's ready to stabilize.

We do think https://github.com/rust-lang/rust/pull/93809 should go in first, and then https://github.com/rust-lang/rust/pull/93587 .

**Comment by Lokathor** [comment]
Was there a discussion regarding the `noreturn` asm option being used on the assembly despite it returning anyway? It has *already* lead at least two people to misunderstand how that asm option works, we should not stabilize that form.

**Comment by Amanieu** [comment]
I just published a proc-macro version of `#[naked]`: https://github.com/Amanieu/naked-function

It works by lowering a naked function to a `global_asm!` block and an `extern "C"` function import.

IMO this covers 99% of the use cases for naked functions.

**Comment by akiekintveld** [comment]
One use case that looks to not (yet?) be covered by that proc-macro is aligned naked functions. This is needed to implement trap/exception/interrupt handlers on some architectures. For example, RISC-V requires direct trap handlers to be aligned to 4 bytes, so I'm currently using this attribute alongside the `fn_align` feature (#82232) to get the desired effect.

**Comment by Amanieu** [comment]
The proc macro can handle various attributes on the function and generate the equivalent assembly code sequence. Currently the following attributes are handled:
- `export_name`
- `no_mangle`
- `link_section`

But it is relatively easy to add support for more attributes such as `repr(align)` or `instruction_set`.

**Comment by bjorn3** [comment]
That proc macro still has the limitation that it is impossible to handle multiple versions of the same crate. Even two different crates will conflict if they happen to use the same function name, despite not using no_mangle.

**Comment by Lokathor** [comment]
@bjorn3 would you say that the only entity able to fix that specific problem is the compiler itself? Would it be fixable if some sort of extra API were added to proc-macros so that they could get the correct full name of the crate they're expanding code for and then use that in some way?

I'm not against having an attribute in the compiler, but if it *can* be just a normal proc-macro that seems convenient from a maintenance stand point. And if it's a proc-macro that has a fairly clear de-sugaring into an existing language feature that's easier to teach. The proc-macro version could presumably eventually be put into `core::arch` or something like that, so it would still be available in a basic program with no deps, same as the attribute form.

**Comment by bjorn3** [comment]
> Would it be fixable if some sort of extra API were added to proc-macros so that they could get the correct full name of the crate they're expanding code for and then use that in some way?

It would need to be the crate name, stable crate id and module path. In addition even that is not enough if we want to support generics.

> The proc-macro version could presumably eventually be put into core::arch or something like that, so it would still be available in a basic program with no deps, same as the attribute form.

The standard library can't use or expose proc macros unless you want to break cross compilation due to the proc macro not being compiled for the right target. (cross compiling crates using rustc_driver is already impossible because of this issue. even cross compiling rustc itself requires the -Zdual-proc-macro hack) All non-decl macros exported by the standard library have to be built into rustc.

**Comment by Lokathor** [comment]
Oh. Well, I guess that moves my vote back to the attribute. It's really gotta be something that works "out of the box", and it definitely has to work with cross compilation.

**Comment by Amanieu** [comment]
> That proc macro still has the limitation that it is impossible to handle multiple versions of the same crate. Even two different crates will conflict if they happen to use the same function name, despite not using no_mangle.

It's somewhat possible to do this by hashing the `CARGO_PKG_VERSION` and `CARGO_PKG_NAME` environment variables into the symbol name. It might required [`proc_macro::tracked_env`](https://doc.rust-lang.org/nightly/proc_macro/tracked_env/index.html) to be stabilized though.

**Comment by bjorn3** [comment]
Even that doesn't work for crates also used by the standard library. They have the same name and version, but a different `-Cmetadata` value and thus different stable crate id.

**Comment by roblabla** [comment]
Honestly, I kind of wish rust had a macro that allowed creating a mangled name out of a string "as if" it was here, e.g. `mangled_name!("test")` would return the same mangled name as if a function/static had been created with that name here.

**Comment by hudson-ayers** [comment]
Is the current status of stabilizing this that someone needs to open and ultimately merge a new/updated version of https://github.com/rust-lang/rust/pull/93809 ? From what I can tell, this was ready for stabilization modulo #93809 , but it fell out of date.

Has this changed? I searched zulip and did not see anything indicating that a different path was preferred. I am trying to figure out whether there is still a path forward to naked functions being stabilized.

**Comment by Amanieu** [comment]
I have reservations about stabilizing naked functions as they are. My main concern is that they are a massive kludge on the codegen size, needing to generate *very* specific LLVM IR to make it work as expected. I would prefer if naked functions were re-implemented on top of `global_asm!`. This could be as a proc macro, like my [naked-function](https://docs.rs/naked-function/latest/naked_function/) crate, or it could be implemented in rustc itself (which would allow for generic naked functions).

**Comment by bradjc** [comment]
> We reviewed this in today's @rust-lang/lang meeting. We'd love to see this stabilized, and we think it's ready to stabilize.
> 
> We do think #93809 should go in first, and then #93587 .

I'm trying to get a sense for the future of stabilizing `naked_functions`. I get some sense that in the last year and a half some users of this feature have gone with alternatives. It's difficult to tell how many users of this feature are interested in stabilizing it. It's also difficult to tell if the sentiment from 2022 is gone to the point where `naked_functions` won't be stabilized.

Is it best to phase out using this feature and projects that use it but want to use stable to find an alternative implementation?

**Comment by Lokathor** [comment]
I want the `naked` attribute basically as is, using the current constrained naked fns functionality.

With the exception that the current part of the design where you put the `noreturn` option on the asm block that does return is the most 100% nonsense part of rust and that needs to be fixed (just change the option name to anything better).

Using global assembly by hand is almost a strictly worse experience for the programmer.

**Comment by Amanieu** [comment]
My main concern is that LLVM's naked function support is a bit of hack. I wouldn't object to stabilizing this as a language feature if the underlying implementation was changed to emit a global_asm internally (at the LLVM level) instead of inline asm inside an empty function.

**Comment by dancrossnyc** [comment]
Having stabilized naked functions would be very helpful for folks who have to do things like write interrupt handlers, system call entry and exit points, and similar things. I have done this in several projects now, and the ability to create such things in a way that still aligns with the visibility and type checking rules of Rust is really valuable; having nicely mangled symbols for such functionality is very nice. It would be fantastic to see this in stable Rust.

That said, I can understand and sympathize with concerns about the implementation. It's kind of a niche case so therefore understandably low priority, but I don't think an effort towards stabilization should be abandoned.

**Comment by haraldh** [comment]
Yeah, I succesfully replaced naked functions with global_asm in my microkernel also. Symbol mangling and nice section setting would have been nice of course.
https://github.com/enarx/enarx/commit/69827cec93a40e972832df7419bbf80072b9e59c

**Comment by kpcyrd** [comment]
I subscribed to this github issue a while ago but also went with global_asm instead:

```rust
global_asm! {
    ".global _start",
    "_start:",
    "mov rdi, rsp",
    "call main"
}
```

Looking into this was related to my blogpost [Writing a Linux executable from scratch with x86_64-unknown-none and Rust](https://vulns.xyz/2023/03/linux-executable-from-scratch-with-x86_64-unknown-none-rust/).

**Comment by berkus** [comment]
I don't really like kludges, and global_asm! seems like a kludge in this case. Would be nice to have working naked functions (also useful for embedded).

> if the underlying implementation was changed to emit a global_asm internally (at the LLVM level) instead of inline asm inside an empty function.

@Amanieu  Can this be emulated at compiler / LLVM level for the moment? (So that we can start writing `#[naked]` and let a future backend update implement it properly)

A cranelift backend can be probably made to emit this properly, but that's a bit far off for general usage.

**Comment by bradjc** [comment]
GIven the lack of clarity on if there is an implementation that is likely to move forward, we're (Tock) [moving](https://github.com/tock/tock/pull/3802) to `global_asm!()`. I suspect it will be easy enough to move back should this get stabilized in the future.

**Comment by Amanieu** [comment]
I have a crate which provides a proc-macro implementation of naked functions based on `global_asm`, which is usable on stable: https://crates.io/crates/naked-function

It has some limitations compared to a proper compiler-based implementation of naked functions, such as lack of support for generic functions and limited support for symbol mangling. However it should support most use cases.

**Comment by dancrossnyc** [comment]
I suppose I would repeat @berkus's question from last week.

As I understand it, the issue with stabilization at the moment is the implementation of naked, and in particular, the way that it generates code (and the annoying `ud2` it tacks on at the end, which is especially aggravating for those of us who are trying to do things like, say, create code that fits in exactly one 4KiB page or something). Is that correct?

If so, does stabilizing this, even with an imperfect implementation, bind us from changing the implementation to be better later? Perhaps to make it in terms of global_asm or something along those lines?

Or am I missing something here?

**Comment by Lokathor** [comment]
Absolutely it makes sense to have a `naked_asm!` macro. Let's do that.

**Comment by folkertdev** [comment]
Over the past days I've been working on some remaining issues for this feature

- https://github.com/rust-lang/rust/pull/127853
- https://github.com/rust-lang/rust/pull/128004

The third component that needs work is how this feature is presented to users.

I think adding `naked_asm!` as suggested above is a good idea, primarily because it is easier to teach than the alternative "like `asm!`, but with a bunch of restrictions". Error messages can also be simpler and more targeted with `naked_asm!`.

So I think the proposal is: the body of a `#[naked]` function must contain one invocation to `naked_asm!`, using normal `asm!` is no longer allowed (but will get a custom error message pointing to `naked_asm!`). Validation of operands will happen as the macro is expanded (rather than in some later pass, which is the current implementation). After expansion `naked_asm!` will be indistinguishable from `asm!` (until codegen, where naked functions take a special path).

## Questions:

- does `naked_asm!` seem like a good idea to explore?
- should `naked_asm!` be allowed outside of `#[naked]` functions? It is just a normal `asm!` with restrictions, so in theory works anywhere `asm!` does. But also it would be kind of weird to see `naked_asm` outside of `#[naked]` functions. 
- should `naked_asm!` require the `options(noreturn)`, or should that be implicit?
- can this macro be added to the standard library under the current `naked_functions` feature, or is some ceremony required?

## Examples

As noted above, we could drop the (currently mandatory) `options(noreturn)`.

```rust
#[naked]
extern "C" fn using_const_generics<const N: u64>(x: u64) -> u64 {
    const M: u64 = 42;

    unsafe {
        naked_asm!(
            "xor rax, rax",
            "add rax, rdi",
            "add rax, {}",
            "add rax, {}",
            "ret",
            const N,
            const M,
            options(noreturn),
        )
    }
}

#[naked]
unsafe extern "C" fn naked_with_args_and_return(a: isize, b: isize) -> isize {
    naked_asm!("lea rax, [rdi + rsi]", "ret", options(noreturn));
}
```

**Comment by chorman0773** [comment]
As a procedural question, I'd like to ask whether the stablization FCP above is still valid, given it's been ~2 years since it finished.

**Comment by traviscross** [comment]
> As a procedural question, I'd like to ask whether the stablization FCP above is still valid, given it's been ~2 years since it finished.

That is an awfully long time ago.  Please `+I-lang-nominate` the PR for stabilization and we'll make a call there about how it should proceed.

**Comment by folkertdev** [comment]
sure

@rustbot label +I-lang-nominated

for context, recent changes mostly implement changes suggested by @Amanieu. The main changes are 

- https://github.com/rust-lang/rust/pull/128138 add an allowlist for what attributes can be used in combination with `#[naked]`. This got some followups for attributes that we cautiously omitted but are actually used and valid in combination with `#[naked]`.
- https://github.com/rust-lang/rust/pull/128004 implement naked functions using global assembly. Versus using LLVM's naked attribute, we get more control, and don't need to do anything special for other backends (e.g. gcc and cranelift are a lot more relevant versus 2 years ago). I'll need some help with integrating this nicely into `rustc_codegen_ssa`.
- https://github.com/rust-lang/rust/pull/128651 add a separate `naked_asm!` macro to generate the global assembly. This can have dedicated documentation, and clear rules. The alternative is to "abuse" `asm!`, but impose a weird mix of restrictions. 

So far this design (still) looks good to me

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

@folkertdev: When I said please nominate the PR for stabilization, what I mean is, please make a pull request that stabilizes this feature (if you're ready to do that) (documentation [here](https://rustc-dev-guide.rust-lang.org/stabilization_guide.html?highlight=stabi#stabilization-pr)).  In that PR, please include a stabilization report.  You can start from the one [here](https://github.com/rust-lang/rust/issues/90957#issuecomment-1028297041) and update it with what has changed since then and why it's ready to stabilize now.  Then please nominate that PR for lang.

**Comment by Lokathor** [comment]
didn't the "naked asm is implemented using global asm" merge like *yesterday*?

I'd love to have this feature, but we should probably give it a little time to bake after a change like that.

**Comment by Qix-** [comment]
It was my understanding that `ret` is not added to the end of a naked function, but instead something akin to a `ud2` or something, and that `options(noreturn)` was forced when using naked functions. Is that no longer the case (or was it never the case, which is likely)?

**Comment by folkertdev** [comment]
> I'd love to have this feature, but we should probably give it a little time to bake after a change like that.

It will have that time, it would still take months to actually reach stable even when the quickest route is taken (and I suspect it won't be).

> It was my understanding that ret is not added to the end of a naked function, but instead something akin to a ud2, and that options(noreturn) was forced when using naked functions. Is that no longer the case (or was it never the case, which is likely)?

the [reference](https://doc.rust-lang.org/reference/inline-assembly.html#r-asm.options.supported-options.noreturn) says:

> noreturn: The asm! block never returns, and its return type is defined as ! (never). Behavior is undefined if execution falls through past the end of the asm code. A noreturn asm block behaves just like a function which doesn’t return; notably, local variables in scope are not dropped before it is invoked.

But naked function often do actually return, they just do it from assembly rather than from rust code. That is a fundamental difference, and we decided that `noreturn` is inaccurate (and just annoying to type). We'd rather document the constraints of `naked_asm!` (one of which is that control flow may not fall through to the end of the function; the assembly must return or diverge) than annotate every call site with an option.

**Comment by Qix-** [comment]
That's not clear to me though. In https://github.com/rust-lang/rust/issues/90957#issuecomment-2539270352 the generated `global_asm!` equivalent shows that the empty naked function has a generated `ret`.

> and we decided that `noreturn` is inaccurate

> the assembly must return or diverge

If the `naked_asm!` block must return or diverge then `noreturn` is implicit and thus an autogenerated `ret` makes no sense / can be harmful. So is the example generated empty function `global_asm!` in the aforementioned comment incorrect, or have the semantics of `#[naked]` functions changed?

**Comment by Lokathor** [comment]
The example in that comment seems fundamentally incomplete, because it didn't have a `naked_asm!` block at all. I don't believe it would compile because you should have exactly one `naked_asm!` use in a `#[naked]` fn.

I agree that the example seems wrong in some form or another.

**Comment by folkertdev** [comment]
that example was incorrect, I've fixed it now. 

Just to be clear: no implicit instructions are ever added to the body: it's just the asm code as written by the user, and the user must make sure that that assembly upholds the contract of `naked_asm!`

**Comment by oberien** [comment]
Would it make sense to have the compiler insert an additional `ud2` after the macro's assembly output as a way to ensure that the `naked_asm` actually returns, or else the program will error during runtime? Without such an `ud2`, if the naked function does not correctly handle its own return, the function lying behind it in the binary will be executed, which could lead to very hard to debug situations. An `ud2` would ease debugging in such cases during development.

**Comment by Lokathor** [comment]
Short answer: no, that is exactly what the compiler must never do (insert *any* instructions that weren't written in the source program).

**Comment by npmccallum** [comment]
@Lokathor I'm not sure I agree. The defined behavior is that the author MUST handle return correctly. Adding a `ud2` at the end is both completely legal and will likely prevent at least one CVE in the future.

Are you worried about being able to predict code size?

**Comment by Lokathor** [comment]
Yes, being able to predict the exact code size is one reason you'd not want `ud2`.

To quote the stabilization report itself:

> The codegen approach was chosen over the llvm naked function attribute because:
> 1) the rust compiler can guarantee the behavior (no sneaky additional instructions, no inlining, etc.)

I think it would be reasonable for *debug* builds, or some opt-in flag, to add the instruction, if you want to "turn on sanitizers". But it should not be the default in --release builds.

**Comment by folkertdev** [comment]
I'm in favor of the simpler policy of never generating additional instructions, but as a datapoint, gcc does actually emit a `ud2` instruction at the end of a naked function (even with optimizations on). Clang does not do this, not even when compiled with no optimizations. 

https://godbolt.org/z/Mr5xY8r31

**Comment by Amanieu** [comment]
I also agree with this: naked_asm is specifically useful when you want *precise* control over what instructions are emitted. We don't want the compiler to insert instructions behind your back.

**Comment by bjorn3** [comment]
Some nop padding will be emitted most of the time anyway without the user having control to keep everything aligned. Might as well replace one nop with an ud2 for some extra safety.

**Comment by Lokathor** [comment]
That very much depends on your platform, doesn't it? On old ARM sections are 4 bytes per a32 instruction and usually aligned to only 4 or maybe 8, so there's *maybe* one `nop` between two different functions, but just as likely there's no natural space and this is increasing the size of things. Further, not all embedded platforms even actually do anything with an illegal instruction, so you could be wasting space for no gain at all.

Again, I'm not against the *possibility* of trap instructions being inserted with a flag or configuration, but it should not be required, and I don't even think it should be default.

**Comment by bjorn3** [comment]
I don't think it should be required, but I think we should do it by default when we can. At least with debug assertions enabled.

**Comment by oberien** [comment]
Summarizing a few brought up points:
* **\+** Adding `ud2` can catch bugs and likely vulnerabilities early during development.
* **\+** The compiler is free to add padding between functions. On x86 (from the [x86 Instruction Set Reference](https://mudongliang.github.io/x86/html/file_module_x86_id_318.html), "Other than raising the invalid opcode exception, this instruction is the same as the NOP instruction."
* **\+** The compiler is free to order functions in any way it sees fit. It could decide to put another function directly after the current one which is just the `ud2` instruction.
* **\+** Adding `ud2` on the modern Intel 64, amd64 and ARM is basically zero-cost. On x86 it's 2 bytes which likely get eliminated (e.g. in the case of `ret; ud2`) in the CPU frontend and don't take up μop cache.
* **\-** Adding `ud2` on older chipsets or in embedded scenarios can be quite costly. For example on the Gameboy Advance arm4t, `undef` is a 4-byte nop, taking up valuable iwram space.
* **\-** Adding `ud2` can require more padding between functions for alignment if the end of a function is at or very close to the alignment boundary.
* `ud2` on some platform might be a simple nop without any actual effect. Generating it will only add cost without any benefit on those platforms.

In my opinion rust strives for safety and security even when using low-level mechanism and unsafe code. Adding `ud2` to places rustc requires execution to never reach adds to this security, allowing possible vulnerabilities to be found earlier. On modern platforms, it's practically zero-cost. However, I do agree that it can cause problems on embedded systems and older chipsets.

Would it be a good compromise to put adding the `ud2` instruction after naked-functions behind the already existing compiler flag `trap-unreachable`? You can already use `-Z trap-unreachable=no` today to make the compiler not emit `ud2` instructions in unreachable places.

**Comment by Qix-** [comment]
@bjorn3 I've not seen `nop` paddings before with naked asm blocks, which architecture are you seeing that on? IMO it's imperative that you get "WYSWYG"-like behavior when writing `naked_asm!{}` as that's sort of the whole point - complete and utter control over the emitted bytecode when it's crucial, especially in embedded and kernel environments.

I also strongly disagree that `ud2`-like instructions should be emitted, both for the reasons already stated, but also, not every case of `naked_asm!{}` is called directly. I have a usecase where `naked_asm!{}` is used during a build step to generate a byte array of instructions used in a kernel environment to bring up secondary cores that are in 16-bit mode, which means paging is disabled and thus instructions have to be written to a direct-mapped page somewhere in lower memory. The kernel is higher-half mapped, which means it resides in index 511 typically, and there's no guarantee that the bytecode is contiguous in rodata nor that it cleanly lives in a single page, thus I have to copy it to contiguous physical memory first before I can boot it.

I could imagine that in similar cases the machine code has to be composed in some non-trivial way, meaning that disparate, perhaps "malformed" (as per the noreturn requirement, for example) `naked_asm!` sections are concatenated to form an otherwise well-defined block of machine code. Silently inserting an opinionated `ud2` in there would wreak havoc on such use cases. Yes, perhaps they could use assemblers for this, but it's a massive plus to be able to do this in Rust, already, without any external CLI tools or extensive build systems.

Incredibly niche use case, I know, but this feature is already squarely in the land of niche. It really would be a huge pain to reverse-engineer the extra `ud2` (nop obviously wouldn't be _as much_ as a problem, but still not ideal), especially on CISC architectures and moreso if it was conditionally emitted based on the build profile (release vs dev).

**Comment by folkertdev** [comment]
An alternative approach: could a lint be designed to warn on possibly invalid naked asm?

I'm not sure how to make this robust, but in principle it would be a `asm_string.ends_with("ret")` sort of check. You could then make that more complicated by checking for jumps and invalid instructions like `ud2` in the final position. It is not robust, and the warning should be explicit about that, but gets users some safety, even for (embedded) targets that would always compile to get the smallest binary size. The lint can just be ignored in cases where it is inaccurate.

**Comment by Qix-** [comment]
That's going down an analysis rabbit hole that isn't really feasible (for this case, at least). Not all textual ASM blocks are going to end in `ret` or something equivalent, and not all generated bycode (binary) blocks will end in that instruction either, even if they're well formed. Barring subroutine blocks at the end (or the usecase I mentioned, where things are being concatenated), on x86 alone there are most of the `jmp` instruction variants, `ret` and `iret`, `sysexit`, etc. - all of which would make the asm block well-formed (as it doesn't fall through). Further, some `jmp` variants _wouldn't_ make it well-formed. For example, in most cases a far JMP might mean the block doesn't return, but in cases where the code selector (CS) has to be changed, you typically far-jump with CS to the next instruction (since `mov cs, 0x08` for example isn't valid for some historical reason; to change CS on x86 you have to far jump with CS).

That level of understanding of bytecode would be an impressive project in its own right, but developing all of that just for linting or automatic `ud2` emission would be astronomically overengineered when in 99% of cases the engineer already knows whether or not they would want to omit `ud2`.

**Comment by bjorn3** [comment]
> I've not seen nop paddings before with naked asm blocks, which architecture are you seeing that on?

On x86_64 there is padding after every function to align the next function to 16 bytes.

> IMO it's imperative that you get "WYSIWYG"-like behavior when writing naked_asm!{} as that's sort of the whole point - complete and utter control over the emitted bytecode when it's crucial, especially in embedded and kernel environments.

Adding ud2 after every naked function is indistinguishable from a function which happens to compile to ud2 being placed right after the naked function, so you still get as much WYSIWYG. WYSIWYG only applies to the function itself, not the padding after it.

> I could imagine that in similar cases the machine code has to be composed in some non-trivial way, meaning that disparate, perhaps "malformed" (as per the noreturn requirement, for example) naked_asm! sections are concatenated to form an otherwise well-defined block of machine code. Silently inserting an opinionated ud2 in there would wreak havoc on such use cases. Yes, perhaps they could use assemblers for this, but it's a massive plus to be able to do this in Rust, already, without any external CLI tools or extensive build systems.

You have to indicate the size of the naked function if you want to copy it's bytes. This size woulf exclude the ud2 instruction inserted by the compiler and thus not be coppied anyway.

**Comment by haraldh** [comment]
Linting? what is valid? ret? iret? some future TDX op code extension to enter a TEE? Please! What are you trying to guard against? The whole asm code is unsafe.

**Comment by folkertdev** [comment]
> That's going down an analysis rabbit hole that isn't really feasible 

yes, fair point. 

I was going to make some argument about most `naked_asm!` blocks ending in `ret` anyway but a quick sampling of github shows that is really not the case. There is a bunch of fun/cursed stuff though https://github.com/search?q=naked_asm%21&type=code&p=1

> Linting? what is valid?

My idea was that it is at least a signal to the programmer to check that their assembly satisfies the constraints: if they deem the lint fired incorrectly, they can just `allow` it, perhaps with some comment explaining the reasoning.

But I agree that it is hard/impossible to actually make it robust (enough) to work in practice.

**Comment by Amanieu** [comment]
> Adding ud2 after every naked function is indistinguishable from a function which happens to compile to ud2 being placed right after the naked function, so you still get as much WYSIWYG. WYSIWYG only applies to the function itself, not the padding after it.

That's not entirely true: ELF symbols have a length and typically any padding bytes are not included in this length since they are inserted by the linker. If we append ud2 then this will observably be different when inspecting the symbol table.

**Comment by bjorn3** [comment]
Given that rustc generates inline assembly which defines the symbol, in the inline assembly rustc can specify the length of the symbol to exclude the ud2 instruction.

**Comment by dancrossnyc** [comment]
Please don't insert UD2's and things like that (or similar things for non-x86 targets) in naked functions. One of the problems with the earlier iterations of naked functions, mentioned above, was that they did this, which made it impossible to have precise control over layout and size.

Here's a motivating example: for example, consider two naked functions placed into a specific linker segment; both need to occupy exactly 4KiB (so, page sized on x86) and further they are both supposed to be aligned on 4K boundaries. One may not care what order they're put into the resulting binary, A before B or B before A, but the size and alignment requirements are absolute.  They're not copied; instead, one relies on the linker to place them correctly. Adding extra UD2's at the end of these the NF's throws this off.

I appreciate the arguments for safety here, and in Rust generally, but naked functions are already incredibly niche: if one is reaching for them one's doing something so far out of the ordinary for most programmers that trying to add a ud2 to catch bugs here feels superfluous.

A response to this may be to use module-level assembly instead of naked functions for this sort of thing. While that's valid, there are some properties of naked functions that make them attractive versus global asm: symbol naming and visibility and so on.

**Comment by Qix-** [comment]
> Might as well replace one nop with an ud2 for some extra safety.

> On x86_64 there is padding after every function to align the next function to 16 bytes.

So in the cases when it's already aligned, there's no extra nop to replace, so no ud2 is emitted. So we're back in the territory of "ud2 sometimes, but not other times, and only in debug mode, and changing the preceding body can affect if a ud2 is there, or not".

So in such a case, it is **functionally impossible to rely on it** and therefore going to masquerade a ton of bugs and cause strange behavior in some cases and not others. Some developers will rely on it being there, erroneously of course. But then be surprised when it's not there. Instead of strictly specifying the safety constraints that must hold, which are there already. Also, nops are there to indicate padding. ud2/similar are not padding instructions. It has side effects, and we **cannot** know how different environments/tools/chips are going to react to them being there, even if they're not hit (e.g. in the case of pipelining).

There is a flimsy at best reason for emitting them, and a multitude of strong reasons _not_ to emit them. Adding in such instructions is subverting expectation for some illusion of safety in a place that is very much opt-in and explicitly, objectively unsafe.

**Comment by Skgland** [comment]
While still worse than a closure, it should at least tidy up the scope, what about defining the function in the static definition  [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=71598d18ebc8a183bbb78ee63c8cbbe8)

**Comment by folkertdev** [comment]
> Unless I'm missing something

I think what you're missing is that naked functions cannot use the rust calling convention: they must always be an `extern "something" fn`. 

Therefore, even if the attribute worked in this position `#[naked] || {}`, the function would still be rejected for using an unsupported calling convention. If we ever get closures that are also `extern fn` somehow, we can then allow the `#[naked]` attribute in that position. For now it is not useful.

**Comment by Qix-** [comment]
Ah yep, I figured I was missing something obvious. Thank you @Skgland and @folkertdev, the workaround works just fine :) Appreciate the info!

**Comment by tgross35** [comment]
With the stabilization at https://github.com/rust-lang/rust/pull/134213, I believe this can be closed.
